R-package/src/xgboost_R.cc

```diff
@@ -136,6 +162,109 @@ SEXP SafeMkChar(const char *c_str, SEXP continuation_token) {
   jconfig["nthread"] = Rf_asInteger(n_threads);
   return Json::Dump(jconfig);
 }
+
+template <class SourceDType, class DestDType>
+void XGCopyTypedArray(const void *src, void *dest, size_t len, size_t stride) {
+  const SourceDType *src_typed = static_cast<const SourceDType*>(src);
+  DestDType *dest_typed = static_cast<DestDType*>(dest);
+  if (stride == 1) {
+    std::copy(src_typed, src_typed + len, dest_typed);
+  } else {
+    for (size_t idx = 0; idx < len; idx++) {
+      dest_typed[idx] = static_cast<DestDType>(src_typed[idx*stride]);
+    }
+  }
+}
+
+template <class SourceDType, class DestDType>
+SEXP XGCopyToRArray(const void *src, size_t len, size_t stride, SEXP continuation_token) {
+  if constexpr (!std::is_same<DestDType, double>::value && !std::is_same<DestDType, int>::value) {
+    LOG(FATAL) << "Internal error: attempting to create R array of unsupported type.";
+  }
+  const bool is_int = std::is_same<DestDType, int>::value;
+  SEXP out = Rf_protect(
+    is_int? SafeAllocInteger(len, continuation_token) : SafeAllocReal(len, continuation_token));
+  void *ptr_out = is_int? static_cast<void*>(INTEGER(out)) : static_cast<void*>(REAL(out));
+  XGCopyTypedArray<SourceDType, DestDType>(src, ptr_out, len, stride);
+  Rf_unprotect(1);
+  return out;
+}
+
+SEXP XGArrayStrToROneDimensional(const char *array_str, SEXP ctoken) {
```

Comment:
We have some code in XGBoost to deal with the array interface `src/data/array_interface.h`.

Reply:
But would it allow simplifying something here? I see it has the same kind of fields so for example there still would need to be a `switch` to make the conversion to R types.

Reply:
Let me give it a try tomorrow.